// material-ui-popup-state/hoverWorkaround.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = hoverWorkaround;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var _withStyles = _interopRequireDefault(require("@material-ui/core/styles/withStyles"));

var styles = {
  _modalRoot: {
    pointerEvents: 'none'
  },
  paper: {
    pointerEvents: 'auto'
  }
};

function hoverWorkaround(Comp) {
  var HoverWorkaround = function HoverWorkaround(_ref) {
    var _ref$classes = _ref.classes,
        _modalRoot = _ref$classes._modalRoot,
        classes = (0, _objectWithoutProperties2.default)(_ref$classes, ["_modalRoot"]),
        ModalClasses = _ref.ModalClasses,
        props = (0, _objectWithoutProperties2.default)(_ref, ["classes", "ModalClasses"]);
    return React.createElement(Comp, (0, _extends2.default)({
      classes: classes,
      ModalClasses: (0, _objectSpread2.default)({}, ModalClasses, {
        root: ModalClasses && ModalClasses.root ? "".concat(ModalClasses.root, " ").concat(_modalRoot) : _modalRoot
      })
    }, props));
  };

  return (0, _withStyles.default)(styles)(HoverWorkaround);
}
